### ISSUE TITLE
ASOF JOIN memory usage

### ISSUE THREAD

**Issue by derekmahar** [duckdb/duckdb#8265]
### What happens?

DuckDB runs out of memory while executing `ASOF JOIN` query in `select_binance_transaction_times_and_prices.sql`.

(Please extract source files from [binance_duckdb_asof_join.zip](https://github.com/duckdb/duckdb/files/12062409/binance_duckdb_asof_join.zip).)

```
$ unzip -l binance_duckdb_asof_join.zip
Archive:  binance_duckdb_asof_join.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
     1043  2023-07-15 13:02   get_binance_trades_to_parquet
      766  2023-07-15 13:10   import_trades_from_parquet.py
      245  2023-07-15 13:21   select_binance_transaction_times_and_prices.sql
      537  2023-07-15 13:19   transaction_times.csv
---------                     -------
     2591                     4 files
```

### To Reproduce

1. Download price history data files for coin pair BTCUSDT:
    ```
    for month in $(seq 1 12)
    do
      # Note: requires bash, duckdb, wget, and GNU zcat.
      ./get_binance_trades_to_parquet BTCUSDT 2022 $month
    done
    ```
1. Import price history for coin pair BTCUSDT into DuckDB database `binance.duckdb`:
    ```
    for month in $(seq 1 12)
    do
      python import_trades_from_parquet.py BTCUSDT 2022 $month
    done
    ```
1. Execute `ASOF JOIN` query:
    ```
    cat select_binance_transaction_times_and_prices.sql | duckdb binance.duckdb
    ```

### OS:

Ubuntu Server 22.04

### DuckDB Version:

v0.8.2-dev1764 07b0b0a2a4

### DuckDB Client:

CLI

### Full Name:

Derek Mahar

### Affiliation:

Self-employed

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by hawkfish** [comment]
Getting some errors:

```sh
zcat: unknown compression format
Error: Invalid Input Error: Error in file "/dev/stdin": CSV options could not be auto-detected. Consider setting parser options manually.

hawkfish@Richards-MacBook-Pro-2 binance_duckdb_asof_join % zcat --version
Apple gzip 403.100.6
```

**Comment by derekmahar** [comment]
Script `get_binance_trades_to_parquet` requires GNU zcat in order to extract the single CSV file from the zip archive and write it to standard output.  I have GNU zcat 1.10 in Ubuntu Server 22.04.2:

```
$ zcat --version
zcat (gzip) 1.10
Copyright (C) 2007, 2011-2018 Free Software Foundation, Inc.
This is free software.  You may redistribute copies of it under the terms of
the GNU General Public License <https://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.

Written by Paul Eggert.
```

My guess is that Apple gzip isn't equivalent to GNU zcat.

Alternatively, if you have BusyBox, temporarily alias `zcat` to `busybox unzip -p`, which in this case behaves the same, or edit script `get_binance_trades_to_parquet` and replace `zcat` with `busybox unzip -p`.  If you have neither of these, in `get_binance_trades_to_parquet`, replace

```
wget $wget_options $file_url |
  zcat - |
  duckdb -c "$copy_sql"
```

with

```
wget --quiet $file_url
unzip -p $coin_pair-trades-$year-$month.zip | duckdb -c "$copy_sql"
```

This will leave both the source zip file and the target Parquet file on your disk, so make sure you have enough disk space to store both.

**Comment by hawkfish** [comment]
OK I finally got the test data built, and according to `top` the peak memory usage is around 30GB. Not great, but also not huge. IF this is more memory than you have available, have you tried using `SET external=True` and setting up a temporary directory?

I am rather appalled by the performance, however:

```
Run Time (s): real 1175.622 user 1713.975005 sys 10745.660641
```

That seems excessive for sorting two columns in a 1.2GR table and making 21 ordered probes into it.

**Comment by derekmahar** [comment]
What is the DuckDB configuration option `external`?  I couldn't find this in the [list of configuration options](https://duckdb.org/docs/sql/configuration).  Must I specify a temporary directory when using a named DuckDB directory?

**Comment by hawkfish** [comment]
Yes you need a temporary directory if you are running out of memory:

```sql
PRAGMA temp_directory='path/to/temp/dir';
```

I did some crude profiling and found that almost all the time (15 minutes) is spent sorting, which seems wrong, so there may be an issue with how AsOf and Window handle sorting without partitioning.

**Comment by derekmahar** [comment]
What do you mean by "without partitioning"?

**Comment by hawkfish** [comment]
> What do you mean by "without partitioning"?

Just that there are no equality conditions. (It's the same code that windowing uses, so sometimes I get the terminology mixed up!) I think the current code creates one partition and sorts it instead of just sorting.

**Comment by derekmahar** [comment]
> Yes you need a temporary directory if you are running out of memory:
> 
> ```sql
> PRAGMA temp_directory='path/to/temp/dir';
> ```

According to the DuckDB documentation, PRAGMA [`temp_directory`](https://duckdb.org/docs/sql/pragmas#temp-directory-for-spilling-data-to-disk--defaults-to-tmp) defaults to `.tmp`.  When I run `cat select_binance_transaction_times_and_prices.sql | duckdb binance.duckdb`, DuckDB creates a directory called `binance.duckdb.tmp`, fills it with many large files, but still runs out of memory and dies.  The virtual machine in which I run the query has 16 GB of memory.

**Comment by derekmahar** [comment]
Limiting memory to 12 GB with `PRAGMA memory_limit="12GB"` doesn't help.

**Comment by hawkfish** [comment]
Yeah I suspect that we have some "unmanaged" memory here. Basically the memory limit doesn't apply to stuff we assume can fit in memory, and it may be that this is what is going on here.  I have your data loaded and I'm currently verifying that the sort code is now parallel but I'll can have a look once that is done.

**Comment by hawkfish** [comment]
This query does require a huge amount of memory right now. When I run it without limited memory it gets up to about 90GB before dying. But I can limit the memory to about 30GB with paging and it takes around 15 minutes to finish.

It does seem to me that for your situation writing your own join is far more tractable, and we should consider trying to detect it. But as far as I can team from profiling, all that is happening is we are just sorting the uncompressed data and it takes a lot of memory (either RAM or disk) to do that.

**Comment by derekmahar** [comment]
Would it be reasonable to require that the right hand table be sorted?  The [ASOF JOIN](https://clickhouse.com/docs/en/sql-reference/statements/select/join#asof-join-usage) operation in Clickhouse makes this assumption.  In this case, the price table is sorted in ascending order by time stamp.

**Comment by hawkfish** [comment]
Unfortunately, we actually don't know that! We are planning to detect and leverage ordering metadata but not yet.

When we get that, we could just have the planner notice that the incoming relation is ordered and use a different physical operator and fall back to the current approach when we need to sort. Then you get the best of both worlds.

I do think there are still some things we can do in your situation because the problem is that your data gets highly compressed (lots of proximity in prices and timestamps) but we uncompress it and copy it and sort it. There is no reason we can't avoid one of the copies and we have work underway to compress intermediate results. Still not as good as knowing the data is already ordered, but it would at least complete.

**Comment by hawkfish** [comment]
@derekmahar I have another change that reduces the memory for your query to about 52GB by avoiding unnecessary partitioning. One thing I notice is that you are representing the prices as doubles, when it looks like they would fit in a `DECIMAL(7,2)`, which is under 32 bits wide. If so, that would reduce the memory requirements by 25%.

**Comment by derekmahar** [comment]
IIRC, I relied on DuckDB's default column data type conversion algorithm to convert the Parquet column data types to DuckDB column types.  Assuming `read_parquet` is similar to `read_csv`, I will specify explicit DuckDB types for each column.

**Comment by derekmahar** [comment]
I deleted my previous comment because it was incorrect.

After changing the data type of price from `DOUBLE` to `DECIMAL(10,2)`, qty from `DOUBLE` to `DECIMAL(12,8)`, and quote_qty from `DOUBLE` to `DECIMAL(16,8)`, I observed that the size of `binance.duckdb` dropped by about 25%.  (I chose a decimal precision of 10 for the price in order to accommodate a relatively large future increase in the Bitcoin price.)

Unfortunately, DuckDB runs out of disk space on my machine when it attempts to run the `ASOF JOIN`.

**Comment by hawkfish** [comment]
@derekmahar Yeah there is another change that is coming that should reduce the memory requirements (at least for your use case) by about 50%. It may not get in for a while, though.

**Comment by derekmahar** [comment]
Poke.

**Comment by hawkfish** [comment]
I just tried a rewrite using `arg_min` and it ran quickly (32s) in about 8GB of RAM:

```sql
CREATE TEMP TABLE transation_times AS
	FROM "fnord.csv";

WITH prices AS (
	SELECT
		time,
		price 
	FROM trades_btcusdt
), candidates AS (
	SELECT
	  UTC_Time,
	  p.price
	FROM transation_times t INNER JOIN prices p
	  ON t.UTC_Time >= p.time
	WHERE t.UTC_Time >= '2022-01-01'
	  AND t.UTC_Time < '2023-01-01'
)
SELECT UTC_Time, arg_min(price, UTC_Time) AS price
FROM candidates
GROUP BY ALL
;
```

The reason this is so effective is that it converts the join into a nested loop join with the small table on the inside. 

```   
┌───────────────────────────┐                             
│       HASH_GROUP_BY       │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             #0            │                             
│      arg_min(#1, #2)      │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│          UTC_Time         │                             
│           price           │                             
│          UTC_Time         │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│          UTC_Time         │                             
│           price           │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│      NESTED_LOOP_JOIN     │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           INNER           │                             
│      time <= UTC_Time     ├──────────────┐              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │              
│           EC: 32          │              │              
└─────────────┬─────────────┘              │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││         SEQ_SCAN          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│       trades_btcusdt      ││      transation_times     │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            time           ││          UTC_Time         │
│           price           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││ Filters: UTC_Time>=2022-01│
│       EC: 1200493891      ││ -01 00:00:00 AND UTC_Time │
│                           ││  <2023-01-01 00:00:00 AND │
│                           ││    UTC_Time IS NOT NULL   │
│                           ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│                           ││           EC: 4           │
└───────────────────────────┘└───────────────────────────┘                             
```
So essentially it just scans the price table and emits candidate matches for each "probe" row, and the aggregate only keeps the best one so far. The "build" scan and aggregation is then completely parallelisable and

The one caveat is that you have to import the `csv` into a temp table so that the optimiser can use the row count statistics  to orient the join.

This is an interesting rewrite idea for the planner and I'll think about it some more.

**Comment by hawkfish** [comment]
Thought about this some more and you can only do this if the only fields from the probe side are join keys. But still seems useful. And if you have equality keys you can use a hash join with a predicate.

**Comment by derekmahar** [comment]
@hawkfish, your [solution using `arg_max`](https://github.com/duckdb/duckdb/issues/8265#issuecomment-1856380441) ran in just under 21 s in an Ubuntu 24.04 virtual machine on my Minisforum UM690 running a Ryzen 9 6900HX.

Your `arg_max` query seems similar to one that I found that uses a [lateral join](https://github.com/timescale/timescaledb-toolkit/issues/162#issuecomment-865216150) to compute an ASOF join in TimescaleDB.  Running an adapted version of my ASOF join query that instead uses a lateral join, DuckDB v1.1.2-dev38 (45559f5eeb) unfortunately runs out of temporary disk space "memory".

In case you are curious, here is the lateral join query that I executed in DuckDB:

```
WITH prices AS (
  SELECT
    time,
    price
  FROM trades_btcusdt
)
SELECT
  UTC_Time,
  p.price
FROM "transaction_times.csv" t,
  LATERAL (
    SELECT price
    FROM prices p
    WHERE p.time <= t.UTC_Time
    ORDER BY p.time DESC
    LIMIT 1
  ) p
WHERE t.UTC_Time >= '2022-01-01'
  AND t.UTC_Time < '2023-01-01'
;
```

So far, ClickHouse is the only database that I've used that has successfully completed my ASOF join query.

**Comment by hawkfish** [comment]
I'm going to have another look at this.

**Comment by hawkfish** [comment]
I have implemented the alternate plan for small probe tables on `main`, so please grab the latest build and see if it works for you. 

There will also be a blog post on this coming out shortly explaining what I did, and the cutoff for the alternate plan is when the probe table is somewhere around 32-64 rows (you have 21). For the regular operator, we should be able to complete it, albeit very slowly due to paging. You will need ~48GB of disk space and a lot of patience. 

In the future we hope to be able to leverage existing ordering metadata to improve the build process, but that will not be soon.

**Comment by derekmahar** [comment]
Thank you for making progress on this issue!

On my Ubuntu 24.04.2 virtual machine having 16 GB of RAM and 214 GB of available disk space, the ASOF join completes in just under 32 s.  What's even more remarkable is that it completes even when I limit the available memory to as low as 45 MB (`PRAGMA memory_limit="45MB"`)!

```
$ duckdb --version
v1.3.0-dev1112 981c7a8573
$ duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb
┌─────────────────────┬───────────────┐
│      UTC_Time       │     price     │
│      timestamp      │ decimal(10,2) │
├─────────────────────┼───────────────┤
│ 2022-07-26 08:53:52 │      21067.47 │
│ 2022-07-26 08:54:21 │      21069.73 │
│ 2022-07-26 22:51:39 │      21113.68 │
│ 2022-11-28 20:02:29 │      16225.90 │
│ 2022-07-12 12:55:35 │      19886.14 │
│ 2022-07-26 09:50:10 │      21140.59 │
│ 2022-07-26 22:52:15 │      21067.18 │
│ 2022-11-27 22:24:55 │      16579.00 │
│ 2022-07-12 10:12:01 │      19694.33 │
│ 2022-07-26 09:50:28 │      21146.11 │
│ 2022-11-27 18:38:37 │      16553.95 │
│ 2022-11-27 18:53:43 │      16552.52 │
│ 2022-07-12 10:10:38 │      19668.07 │
│ 2022-07-26 09:50:15 │      21145.03 │
│ 2022-11-28 14:10:02 │      16179.06 │
│ 2022-07-12 10:10:28 │      19666.58 │
│ 2022-12-31 21:52:37 │      16549.60 │
│ 2022-07-12 10:10:43 │      19670.73 │
│ 2022-07-26 08:51:51 │      21062.34 │
│ 2022-07-26 22:51:05 │      21115.89 │
│ 2022-11-29 20:08:39 │      16490.01 │
├─────────────────────┴───────────────┤
│ 21 rows                   2 columns │
└─────────────────────────────────────┘
```

Using [Hyperfine](https://github.com/sharkdp/hyperfine) to measure query execution time:

```
$ hyperfine --warmup 3 "duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb"
Benchmark 1: duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb
Time (mean ± σ):     31.852 s ±  0.464 s    [User: 202.831 s, System: 17.235 s]
Range (min … max):   30.969 s … 32.417 s    10 runs
```

**Comment by derekmahar** [comment]
Unfortunately, when I use a version of `transactions_times.csv` having 240 data rows (by duplicating the same original 24 data rows ten times), DuckDB CLI dies after quickly running out of memory while executing the ASOF join.

**Comment by derekmahar** [comment]
After setting the `memory_limit` to an arbitrary 5 GB and `temp_directory` to the directory in which I ran DuckDB CLI, DuckDB CLI executed the query in an average of 1611 s.

```
SET memory_limit="5GB";
SET temp_directory='.';
```

Time to execute the query:
```
$ hyperfine "duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb"
Benchmark 1: duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb
  Time (mean ± σ):     1610.677 s ± 79.426 s    [User: 1054.357 s, System: 4864.891 s]
  Range (min … max):   1388.602 s … 1654.972 s    10 runs
```

**Comment by hawkfish** [comment]
The improvement I made was to plan a nested loop join when the probe table is small. Unfortunately, that is a quadratic operation, so it gets Really Bad after the probe side gets above 40 rows or so. I wrote a [blog post](https://duckdb.org/2025/02/19/asof-plans.html) on the whole saga if you want the gory details. But it's pretty clear from the measurements that 240 rows is way above the performance cutoff.

Most of the memory is spent sorting the data and we have improvements to the sort code coming.  There is also an issue related to some scheduling work I did on the window operator where we are not using multiple cores for a single partition/equality condition. This is especially bad for AsOf when there is only an inequality (like windowing without partitions). Fixing this will not improve the sort memory footprint but should allow greater parallelisation of the probe phase.  Even better, the two changes will interact because the new sort code should allow me to scan the sorted data at a much finer granularity, enabling higher parallelism.

**Comment by derekmahar** [comment]
Would it not improve query performance considerably if the query command could inform the query planner that the right-hand side table (the "lookup" or "reference" table) is already sorted so that the planner could skip the sorting operation?

This scenario would be common in time series cases like mine where the events in the reference table are historical or archival, will never change, and might already be sorted by timestamp (perhaps "frozen" in read-only Parquet files) or must be sorted only once.  This scenario further assumes that any new events added to the time series archive (for example, in a new Parquet file for the subsequent period) are greater in time than earlier events already in the archive and are in ascending time order.

**Comment by derekmahar** [comment]
Would it not improve query performance considerably if the query command could inform the query planner that the right-hand side table (the "lookup" or "reference" table) is already sorted so that the planner could skip the sorting operation?

This scenario would be common in time series cases like mine where the events in the reference table are historical or archival, will never change, and might already be sorted by timestamp (perhaps "frozen" in read-only Parquet files) or must be sorted only once.  This scenario further assumes that any new events added to the time series archive (for example, in a new Parquet file for the subsequent period) are greater in time than earlier events already in the archive and are in ascending time order.  (In this scenario, it would be useful if  one could specify in the query the order in which to read the input Parquet files.)

**Comment by hawkfish** [comment]
I think you are suggesting a "hint"? We don't currently have a hinting system, but [detecting, tracking and leveraging ordering metadata](https://github.com/duckdblabs/duckdb-internal/issues/3741) is something we are looking into.

**Comment by derekmahar** [comment]
> I think you are suggesting a "hint"? We don't currently have a hinting system, but [detecting, tracking and leveraging ordering metadata](https://github.com/duckdblabs/duckdb-internal/issues/3741) is something we are looking into.

Yes, this a hinting system would be very useful. Unfortunately, I can't view the reference that you posted because it appears to refer to a private repository.

**Comment by hawkfish** [comment]
@derekmahar Can you try this with the latest 1.5 build? I've done a lot of work on AsOf recently, both to increase parallelism and reduce memory usage.

**Comment by hawkfish** [comment]
I just ran this with the code on `main` with the "small probe" optimisation disabled. Peak memory usage was ~28GB and the run time was 113s on my MacBook Pro. The query now uses all 8 cores (hyper threading doesn't seem to help).

```sql
load '~/Documents/Data/binance_duckdb_asof_join/binance.duckdb' readonly

PRAGMA asof_loop_join_threshold=0;

WITH prices AS (
  SELECT
    time,
    price 
  FROM trades_btcusdt
)
SELECT
  UTC_Time,
  p.price
FROM probe t
  ASOF JOIN prices p
  ON t.UTC_Time >= p.time
WHERE t.UTC_Time >= '2022-01-01'::TIMESTAMP
  AND t.UTC_Time < '2023-01-01'::TIMESTAMP
;
```

|      UTC_Time       |  price   |
|---------------------|---------:|
| 2022-07-12 10:10:28 | 19666.58 |
| 2022-07-12 10:10:38 | 19668.07 |
| 2022-07-12 10:10:43 | 19670.73 |
| 2022-07-12 10:12:01 | 19694.33 |
| 2022-07-12 12:55:35 | 19886.4  |
| 2022-07-26 08:51:51 | 21062.34 |
| 2022-07-26 08:53:52 | 21067.47 |
| 2022-07-26 08:54:21 | 21069.73 |
| 2022-07-26 09:50:10 | 21140.62 |
| 2022-07-26 09:50:15 | 21145.03 |
| 2022-07-26 09:50:28 | 21146.11 |
| 2022-07-26 22:51:05 | 21115.89 |
| 2022-07-26 22:51:39 | 21113.68 |
| 2022-07-26 22:52:15 | 21067.04 |
| 2022-11-27 18:38:37 | 16553.94 |
| 2022-11-27 18:53:43 | 16552.52 |
| 2022-11-27 22:24:55 | 16579.0  |
| 2022-11-28 14:10:02 | 16179.06 |
| 2022-11-28 20:02:29 | 16225.9  |
| 2022-11-29 20:08:39 | 16487.77 |
| 2022-12-31 21:52:37 | 16549.61 |

**Comment by hawkfish** [comment]
Reducing the memory limit to 12GB runs to completion in a similar time with no OOM.

**Comment by derekmahar** [comment]
@hawkfish This is a remarkable performance improvement!

For the 240 transaction "probe" table (`transaction_times_2.csv`), I saw no out-of-memory exceptions and much faster average execution time than what I recorded [previously](https://github.com/duckdb/duckdb/issues/8265#issuecomment-2709022020) :

```
$ duckdb --version
v1.5.0-dev2458 (Development Version) 39f5583f99

# Benchmark using transaction_times_2.csv (240 transactions, same 24 transactions in transaction_times.csv concatenated 10 times)
$ hyperfine --warmup 3 "duckdb -f select_binance_transaction_times_and_prices_2.sql binance.duckdb"
Benchmark 1: duckdb -f select_binance_transaction_times_and_prices_2.sql binance.duckdb
  Time (mean ± σ):     87.907 s ±  3.486 s    [User: 294.564 s, System: 135.808 s]
  Range (min … max):   83.441 s … 94.033 s    10 runs
```

```sql
# select_binance_transaction_times_and_prices_2.sql
WITH prices AS (
  SELECT
    time,
    price
  FROM trades_btcusdt
),
transaction_times AS (
  SELECT *
  FROM "transaction_times_2.csv"
  ORDER BY UTC_Time
)
SELECT
  UTC_Time,
  p.price
FROM transaction_times t
  ASOF JOIN prices p
  ON t.UTC_Time >= p.time
WHERE t.UTC_Time >= '2022-01-01'
  AND t.UTC_Time < '2023-01-01'
ORDER BY UTC_Time
;
```
[select_binance_transaction_times_and_prices_2.sql](https://github.com/user-attachments/files/23506323/select_binance_transaction_times_and_prices_2.sql)
[transaction_times_2.csv](https://github.com/user-attachments/files/23506289/transaction_times_2.csv)

The average execution time using the 24 transaction probe table (`transaction_times.csv`) was comparable to what I recorded [previously](https://github.com/duckdb/duckdb/issues/8265#issuecomment-2708808240):

```
# Benchmark using transaction_times.csv (24 transactions)
$ hyperfine --warmup 3 "duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb"
Benchmark 1: duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb
  Time (mean ± σ):     32.282 s ±  0.897 s    [User: 244.708 s, System: 7.624 s]
  Range (min … max):   31.607 s … 34.601 s    10 runs
```

```sql
# select_binance_transaction_times_and_prices.sql
WITH prices AS (
  SELECT
    time,
    price
  FROM trades_btcusdt
),
transaction_times AS (
  SELECT *
  FROM "transaction_times.csv"
  ORDER BY UTC_Time
)
SELECT
  UTC_Time,
  p.price
FROM transaction_times t
  ASOF JOIN prices p
  ON t.UTC_Time >= p.time
WHERE t.UTC_Time >= '2022-01-01'
  AND t.UTC_Time < '2023-01-01'
ORDER BY UTC_Time
;
```
[select_binance_transaction_times_and_prices.sql](https://github.com/user-attachments/files/23506322/select_binance_transaction_times_and_prices.sql)
[transaction_times.csv](https://github.com/user-attachments/files/23506288/transaction_times.csv)

For several transaction times, query `select_binance_transaction_times_and_prices.sql` (which uses probe table `transaction_times.csv`) found different prices than those which your query found:

```
$ duckdb -f select_binance_transaction_times_and_prices.sql binance.duckdb > binance_transaction_times_and_prices_1.csv
$ diff binance_transaction_times_and_prices_1.csv binance_transaction_times_and_prices_hawkfish.csv
6c6
< 2022-07-12 12:55:35,19886.14
---
> 2022-07-12 12:55:35,19886.40
10c10
< 2022-07-26 09:50:10,21140.59
---
> 2022-07-26 09:50:10,21140.62
15,16c15,16
< 2022-07-26 22:52:15,21067.18
< 2022-11-27 18:38:37,16553.95
---
> 2022-07-26 22:52:15,21067.04
> 2022-11-27 18:38:37,16553.94
21,22c21,22
< 2022-11-29 20:08:39,16490.01
< 2022-12-31 21:52:37,16549.60
---
> 2022-11-29 20:08:39,16487.77
> 2022-12-31 21:52:37,16549.61
```
[binance_transaction_times_and_prices_1.csv](https://github.com/user-attachments/files/23507028/binance_transaction_times_and_prices_1.csv)
[binance_transaction_times_and_prices_hawkfish.csv](https://github.com/user-attachments/files/23507029/binance_transaction_times_and_prices_hawkfish.csv)

The prices for at least the last two transaction times that query `select_binance_transaction_times_and_prices.sql` selected appear to be correct:

```
$ duckdb binance.duckdb "SELECT trade_id, time, price FROM trades_btcusdt WHERE trade_id = (SELECT arg_max(t
rade_id, time) FROM trades_btcusdt WHERE time <= '2022-11-29 20:08:39')"
┌────────────┬─────────────────────────┬───────────────┐
│  trade_id  │          time           │     price     │
│   int64    │        timestamp        │ decimal(10,2) │
├────────────┼─────────────────────────┼───────────────┤
│ 2263805772 │ 2022-11-29 20:08:38.995 │      16490.01 │
└────────────┴─────────────────────────┴───────────────┘
$ duckdb -c "SELECT trade_id, epoch_ms(time), price FROM '../BTCUSDT/2022/BTCUSDT-trades-2022-11.parquet' WH
ERE trade_id = (SELECT arg_max(trade_id, time) FROM '../BTCUSDT/2022/BTCUSDT-trades-2022-11.parquet' WHERE epoch_ms(time) <= '2022-11-29 20:08:39')"
┌────────────┬─────────────────────────┬────────────────┐
│  trade_id  │    epoch_ms("time")     │     price      │
│   int64    │        timestamp        │ decimal(20,8)  │
├────────────┼─────────────────────────┼────────────────┤
│ 2263805772 │ 2022-11-29 20:08:38.995 │ 16490.01000000 │
└────────────┴─────────────────────────┴────────────────┘
$ duckdb -c "SELECT trade_id, epoch_ms(time) AS time, price FROM read_csv('BTCUSDT-trades-2022-12.csv', columns = {trade_id: BIGINT, price: DECIMAL, qty: DECIMAL, quote_qty: DECIMAL, time: BIGINT, is_buyer_maker: BOOLEAN, is_buyer_match: BOOLEAN}) WHERE epoch_ms(time) <= '2022-12-31 21:52:37' ORDER BY epoch_ms(time) DESC LIMIT 1"
┌────────────┬─────────────────────────┬───────────────┐
│  trade_id  │          time           │     price     │
│   int64    │        timestamp        │ decimal(18,3) │
├────────────┼─────────────────────────┼───────────────┤
│ 2407865351 │ 2022-12-31 21:52:36.987 │     16549.600 │
└────────────┴─────────────────────────┴───────────────┘
$ duckdb binance.duckdb "SELECT trade_id, time, price FROM trades_btcusdt WHERE trade_id = (SELECT arg_max(t
rade_id, time) FROM trades_btcusdt WHERE time <= '2022-12-31 21:52:37')"
┌────────────┬─────────────────────────┬───────────────┐
│  trade_id  │          time           │     price     │
│   int64    │        timestamp        │ decimal(10,2) │
├────────────┼─────────────────────────┼───────────────┤
│ 2407865351 │ 2022-12-31 21:52:36.987 │      16549.60 │
└────────────┴─────────────────────────┴───────────────┘
$ duckdb -c "SELECT trade_id, epoch_ms(time), price FROM '../BTCUSDT/2022/BTCUSDT-trades-2022-12.parquet' WH
ERE trade_id = (SELECT arg_max(trade_id, time) FROM '../BTCUSDT/2022/BTCUSDT-trades-2022-12.parquet' WHERE epoch_ms(time) <= '2022-12-31 21:52:37')"
┌────────────┬─────────────────────────┬────────────────┐
│  trade_id  │    epoch_ms("time")     │     price      │
│   int64    │        timestamp        │ decimal(20,8)  │
├────────────┼─────────────────────────┼────────────────┤
│ 2407865351 │ 2022-12-31 21:52:36.987 │ 16549.60000000 │
└────────────┴─────────────────────────┴────────────────┘
$ duckdb -c "SELECT trade_id, epoch_ms(time) AS time, price FROM read_csv('BTCUSDT-trades-2022-12.csv', columns = {trade_id: BIGINT, price: DECIMAL, qty: DECIMAL, quote_qty: DECIMAL, time: BIGINT, is_buyer_maker: BOOLEAN, is_buyer_match: BOOLEAN}) WHERE epoch_ms(time) <= '2022-12-31 21:52:37' ORDER BY epoch_ms(time) DESC LIMIT 1"
┌────────────┬─────────────────────────┬───────────────┐
│  trade_id  │          time           │     price     │
│   int64    │        timestamp        │ decimal(18,3) │
├────────────┼─────────────────────────┼───────────────┤
│ 2407865351 │ 2022-12-31 21:52:36.987 │     16549.600 │
└────────────┴─────────────────────────┴───────────────┘
```

Query `select_binance_transaction_times_and_prices_2.sql` (which uses probe table `transaction_times_2.csv`) also selected some incorrect prices that were the same as those which your query selected:

```
$ duckdb -csv -f select_binance_transaction_times_and_prices_2.sql binance.duckdb > binance_transaction_times_and_prices_2.csv
$ tail -n 30 binance_transaction_times_and_prices_2.csv | uniq
2022-11-28 20:02:29,16225.90
2022-11-29 20:08:39,16487.77
2022-12-31 21:52:37,16549.61
```
[binance_transaction_times_and_prices_2.csv](https://github.com/user-attachments/files/23507027/binance_transaction_times_and_prices_2.csv)

**Comment by hawkfish** [comment]
Being able to scan the sorted data in parallel at high granularity helps a lot.

If there is a correctness issue, could you file a new targeted issue?
